arrow-fx-stm/src/main/kotlin/arrow/fx/stm/internal/Hamt.kt

```diff
@@ -0,0 +1,146 @@
+  while (true) {
+    val branchInd = hash.indexAtDepth(depth)
+    val branches = hmt.branches.read()
+    when (val branch = branches[branchInd]) {
+      null -> {
+        val el = fn(null) ?: return false
+        val new = branches.copyOf()
+        new[branchInd] = Branch.Leaf(hash, arrayOf(el))
+        hmt.branches.write(new)
+        return true
+      }
+      is Branch.Leaf -> {
+        if (hash == branch.hash) {
+          val ind = branch.value.indexOfFirst(test).takeIf { it != -1 }
+          when (val el = ind?.let { branch.value[it] }) {
+            null -> {
+              // insert new value with a colliding hash
+              val newEl = fn(null) ?: return false
+              val new = branches.copyOf()
+              new[branchInd] = Branch.Leaf(hash, arrayOf(newEl, *branch.value))
+              hmt.branches.write(new)
+              return true
+            }
+            else -> {
+              when (val newEl = fn(el)) {
+                null -> {
+                  // remove element
+                  if (branch.value.size > 1) {
+                    val newLeafArray = Array(branch.value.size - 1) { i ->
+                      if (i >= ind) branch.value[i + 1]
+                      else branch.value[i]
+                    }
+                  } else {
+                    val new = branches.copyOf()
+                    new[branchInd] = null
+                    hmt.branches.write(new)
+                  }
+                  return true
+                }
+                else -> {
+                  // update element
+                  val newLeafArr = branch.value.copyOf()
+                  newLeafArr[ind] = newEl
+                  val new = branches.copyOf()
+                  new[branchInd] = Branch.Leaf(hash, newLeafArr)
+                  hmt.branches.write(new)
+                  return true
+                }
+              }
+            }
+          }
+        } else {
+          val el = fn(null) ?: return false
+          val newHamt = pair(
+            depth.nextDepth(),
+            hash,
+            Branch.Leaf(hash, arrayOf(el)),
+            branch.hash,
+            branch
+          )
+          val new = branches.copyOf()
+          new[branchInd] = Branch.Branches(newHamt)
+          hmt.branches.write(new)
+          return true
+        }
+      }
+      is Branch.Branches -> {
+        depth = depth.nextDepth()
+        hmt = branch.sub
+      }
+    }
+  }
+}
+
+fun <A> STM.newHamt(): Hamt<A> = Hamt(newTVar(arrayOfNulls(ARR_SIZE)))
+
+sealed class Branch<out A> {
+  data class Branches<A>(val sub: Hamt<A>) : Branch<A>()
+  data class Leaf<A>(val hash: Int, val value: Array<A>) : Branch<A>()
+}
+
+const val ARR_SIZE = 32 // 2^DEPTH_STEP
+const val DEPTH_STEP = 5
+const val MASK = 1.shl(DEPTH_STEP) - 1
+
+fun Int.index(): Int = MASK.and(this)
+fun Int.atDepth(d: Int): Int = shr(d)
+fun Int.indexAtDepth(d: Int): Int = atDepth(d).index()
+fun Int.nextDepth(): Int = this + DEPTH_STEP
```

Comment:
I believe this was due to usage of inline somewhere above, I'll check tomorrow if I can lower that...

Reply:
Thank you @1Jajen1!

Reply:
Ah yes this is because `alterHamtWithHash` is public and inline. It has to be inline because it uses array creation methods from the stdlib which use reified afaik. This is really annoying :/

Reply:
Then let's merge :)